kv/notification_endpoint.go

```diff
@@ -0,0 +1,544 @@
+		Msg:  "provided notification endpoint ID has invalid format",
+	}
+)
+
+var _ influxdb.NotificationEndpointService = (*Service)(nil)
+
+func (s *Service) initializeNotificationEndpoint(ctx context.Context, tx Tx) error {
+	if _, err := s.notificationEndpointBucket(tx); err != nil {
+		return err
+	}
+	if _, err := s.notificationEndpointIndexBucket(tx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// UnavailableNotificationEndpointStoreError is used if we aren't able to interact with the
+// store, it means the store is not available at the moment (e.g. network).
+func UnavailableNotificationEndpointStoreError(err error) *influxdb.Error {
+	return &influxdb.Error{
+		Code: influxdb.EInternal,
+		Msg:  fmt.Sprintf("Unable to connect to notification endpoint store service. Please try again; Err: %v", err),
+		Op:   "kv/notificationEndpoint",
+	}
+}
+
+// UnavailableNotificationEndpointIndexError is used when the error comes from an internal system.
+func UnavailableNotificationEndpointIndexError(err error) *influxdb.Error {
+	return &influxdb.Error{
+		Code: influxdb.EInternal,
+		Msg:  fmt.Sprintf("unexpected error retrieving notfication endpoint's index bucket; Err %v", err),
+		Op:   "kv/notificationEndpointIndex",
+	}
+}
+
+// InternalNotificationEndpointStoreError is used when the error comes from an
+// internal system.
+func InternalNotificationEndpointStoreError(err error) *influxdb.Error {
+	return &influxdb.Error{
+		Code: influxdb.EInternal,
+		Msg:  fmt.Sprintf("Unknown internal notificationEndpoint data error; Err: %v", err),
+		Op:   "kv/notificationEndpoint",
+	}
+}
+
+func (s *Service) notificationEndpointBucket(tx Tx) (Bucket, error) {
+	b, err := tx.Bucket(notificationEndpointBucket)
+	if err != nil {
+		return nil, UnavailableNotificationEndpointStoreError(err)
+	}
+	return b, nil
+}
+
+func (s *Service) notificationEndpointIndexBucket(tx Tx) (Bucket, error) {
+	b, err := tx.Bucket(notificationEndpointIndex)
+	if err != nil {
+		return nil, UnavailableNotificationEndpointIndexError(err)
+	}
+	return b, nil
+}
+
+// CreateNotificationEndpoint creates a new notification endpoint and sets b.ID with the new identifier.
+func (s *Service) CreateNotificationEndpoint(ctx context.Context, edp influxdb.NotificationEndpoint, userID influxdb.ID) error {
+	return s.kv.Update(ctx, func(tx Tx) error {
+		return s.createNotificationEndpoint(ctx, tx, edp, userID)
+	})
+}
+
+func (s *Service) createNotificationEndpoint(ctx context.Context, tx Tx, edp influxdb.NotificationEndpoint, userID influxdb.ID) error {
+	if edp.GetOrgID().Valid() {
+		span, ctx := tracing.StartSpanFromContext(ctx)
+		defer span.Finish()
+
+		_, pe := s.findOrganizationByID(ctx, tx, edp.GetOrgID())
+		if pe != nil {
+			return &influxdb.Error{
+				Op:  influxdb.OpCreateCheck,
+				Err: pe,
+			}
+		}
+	}
+	// notification endpoint name unique
+	if _, err := s.findNotificationEndpointByName(ctx, tx, edp.GetOrgID(), edp.GetName()); err == nil {
```

Comment:
Is it necessary to enforce uniqueness by name? We don't have a name param in the API.

Reply:
I think it's fine to make the name, since the dropdown of same name endpoints could confuse user.